### PR TITLE
Add test to record order of evaluation.

### PR DESCRIPTION
**Pull request by gafter**

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs [hunk 1]
```diff
@@ -2960,5 +2960,75 @@ public class MatchFailureException : InvalidOperationException
             CompileAndVerify(compilation, expectedOutput: @"3
 MatchFailureException()");
         }
+
+        [Fact]
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs]: >Fact [](start = 9, length = 4)

nit: note that the current formatter inserts a space between the type and the positional sub-patterns: `A (...`.
You may want to adjust the formatting in these tests.
For reference, `FormatRecursivePattern_SpaceBetweenTypeAndPositionalSubpattern`

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs [hunk 2]
```diff
@@ -2960,5 +2960,75 @@ public class MatchFailureException : InvalidOperationException
...
+@"A(A(1, A(2, 3)), A(4, A(5, 6))).Deconstruct
+A(1, A(2, 3)).Deconstruct
+A(2, 3).Deconstruct
+A(2, 3).X
+A(4, A(5, 6)).Deconstruct
+A(5, 6).Deconstruct
+A(5, 6).Y
+A(2, 3).Y
+8");
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs]: > } [](start = 7, length = 2)

Is it fair to say that any order of evaluation verified here is "correct" (the spec will not commit to a specific order of evaluation), as long as we don't have double-evaluations?

Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs]: Almost. Double evaluations are permitted too.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs]: I will report a bug against the formatter.  There is a design principle of patterns that they have a certain correspondence to construction syntax, and you would not write `new A (1)` to create, so you should not write `A (1)` to match.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs [hunk 3]
```diff
@@ -2960,5 +2960,75 @@ public class MatchFailureException : InvalidOperationException
             CompileAndVerify(compilation, expectedOutput: @"3
 MatchFailureException()");
         }
+
+        [Fact]
+        public void RecordOrderOfEvaluation()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs]: Do you want to add a similar test capturing the order of evaluation of property getters involved in evaluating a property pattern?

Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs]: Those are in this test.

**Pull request merged**